gnucash/gnome-utils/gnc-autoclear.cpp

```diff
@@ -0,0 +1,281 @@
+ * published by the Free Software Foundation; either version 2 of   *
+ * the License, or (at your option) any later version.              *
+ *                                                                  *
+ * This program is distributed in the hope that it will be useful,  *
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of   *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the    *
+ * GNU General Public License for more details.                     *
+ *                                                                  *
+ * You should have received a copy of the GNU General Public License*
+ * along with this program; if not, contact:                        *
+ *                                                                  *
+ * Free Software Foundation           Voice:  +1-617-542-5942       *
+ * 51 Franklin Street, Fifth Floor    Fax:    +1-617-542-2652       *
+ * Boston, MA  02110-1301,  USA       gnu@gnu.org                   *
+ *******************************************************************/
+#include <config.h>
+
+#include <gtk/gtk.h>
+#include <glib/gi18n.h>
+#include <vector>
+#include <unordered_map>
+
+#include "inttypes.h"
+#include "Account.h"
+#include "Split.h"
+#include "Transaction.h"
+#include "gncOwner.h"
+#include "qof.h"
+#include "gnc-autoclear.h"
+
+static QofLogModule log_module = GNC_MOD_GUI;
+
+/* the following functions are used in window-autoclear: */
+
+typedef enum
+{
+    AUTOCLEAR_OVERLOAD = 1,
+    AUTOCLEAR_UNABLE,
+    AUTOCLEAR_MULTIPLE,
+    AUTOCLEAR_NOP,
+} autoclear_error_type;
+
+#define MAX_AUTOCLEAR_SECONDS 10
+
+#define MAX_AUTOCLEAR_TICKS CLOCKS_PER_SEC * MAX_AUTOCLEAR_SECONDS
+
+using SplitVec = std::vector<Split*>;
+
+struct WorkItem
+{
+    gint64 reachable_amount;
+    SplitVec splits_vector;
+    WorkItem (const gint64 amount, SplitVec&& splits)
+        : reachable_amount (amount), splits_vector(std::move(splits)){}
+};
+
+static void status_update (GtkLabel *label, gchar *status)
+{
+    if (!label) return;
+    gtk_label_set_text (label, status);
+    while (gtk_events_pending ())
+        g_main_context_iteration (nullptr, FALSE);
+}
+
+static void looping_update_status (GtkLabel *label, guint nc_progress,
+                                   guint nc_size, guint size)
+{
+    if (!label)
+        return;
+
+    auto text = g_strdup_printf ("%u/%u splits processed, %u combos",
+                                       nc_progress, nc_size, size);
+    status_update (label, text);
+    g_free (text);
+}
+
+class Hasher
+{
+public:
+    size_t operator() (gint64 const& key) const
+    {
+        return key;
+    }
+};
+class EqualFn
+{
+public:
+    bool operator() (gint64 const& t1, gint64 const& t2) const
+    {
+        return (t1 == t2);
+    }
+};
+
+static inline gint64
+normalize_num (gnc_numeric amount, int first_denom)
+{
+    return ((amount.num * first_denom) / amount.denom);
+};
+
```

Comment:
The concern here is that we need to account for large denoms which could overflow. This function is called infrequently (nc_vector_size() + 1) therefore doesn't need to be too fast.

Reply:
```cpp
static inline gint64
normalize_num (gnc_numeric amount, int first_denom)
{
    auto cmp = first_denom - amount.denom;
    if (G_LIKELY (cmp == 0))
        return amount.num;
    else if (cmp > 0)       // e.g. gnc_numeric is 1230/10 and first_denom is 100
    {
        auto ratio = first_denom / amount.denom;
        return amount.num * ratio;
    }
    else                  // e.g. gnc_numeric is 123000/1000 and first_denom is 100
    {
        auto ratio = amount.denom / first_denom;
        return amount.num / ratio;
    }
};
```

Reply:
This is very unsafe against overflow if the denominators are large, especially if they don't share common factors. You also risk getting wedged where a total of several splits is no longer convertible to the denominator of toclear_value

Instead use  [gnc_numeric_convert](https://github.com/Gnucash/gnucash/blob/aa450cb00eb9c10d07113fe36ec992b684f981c0/libgnucash/engine/gnc-numeric.h#L457); it should always be possible and safe to convert to the commodity's SCU, after which you can safely use the numerator for all computation.